File: android/app/src/main/java/my/ssdid/mobile/domain/transport/ServerApi.kt
package my.ssdid.mobile.domain.transport

import my.ssdid.mobile.domain.transport.dto.*
import retrofit2.http.Body
import retrofit2.http.POST

interface ServerApi {
    @POST("api/register")
    suspend fun registerStart(@Body request: RegisterStartRequest): RegisterStartResponse

    @POST("api/register/verify")
    suspend fun registerVerify(@Body request: RegisterVerifyRequest): RegisterVerifyResponse

    @POST("api/authenticate")
    suspend fun authenticate(@Body request: AuthenticateRequest): AuthenticateResponse

    @POST("api/transaction/challenge")
    suspend fun requestChallenge(@Body request: TxChallengeRequest): TxChallengeResponse

    @POST("api/transaction/submit")
    suspend fun submitTransaction(@Body request: TxSubmitRequest): TxSubmitResponse
}

File: ios/SsdidWallet/Domain/Transport/ServerApi.swift
import Foundation

/// Service/Server API client for registration, authentication, and transaction signing.
final class ServerApi {

    private let client: SsdidHttpClient
    private let baseURL: String

    init(client: SsdidHttpClient, baseURL: String) {
        self.client = client
        self.baseURL = baseURL
    }

    // MARK: - Registration

    func registerStart(request: RegisterStartRequest) async throws -> RegisterStartResponse {
        return try await client.post(
            url: "\(baseURL)/api/register",
            body: request,
            responseType: RegisterStartResponse.self
        )
    }

    func registerVerify(request: RegisterVerifyRequest) async throws -> RegisterVerifyResponse {
        return try await client.post(
            url: "\(baseURL)/api/register/verify",
            body: request,
            responseType: RegisterVerifyResponse.self
        )
    }

    // MARK: - Authentication

    func authenticate(request: AuthenticateRequest) async throws -> AuthenticateResponse {
        return try await client.post(
            url: "\(baseURL)/api/authenticate",
            body: request,
            responseType: AuthenticateResponse.self
        )
    }

    func getAuthChallenge() async throws -> AuthChallengeResponse {
        return try await client.get(
            url: "\(baseURL)/api/auth/challenge",
            responseType: AuthChallengeResponse.self
        )
    }

    func verifyAuth(request: AuthVerifyRequest) async throws -> AuthVerifyResponse {
        return try await client.post(
            url: "\(baseURL)/api/auth/verify",
            body: request,
            responseType: AuthVerifyResponse.self
        )
    }

    // MARK: - Transaction

    func requestChallenge(request: TxChallengeRequest) async throws -> TxChallengeResponse {
        return try await client.post(
            url: "\(baseURL)/api/transaction/challenge",
            body: request,
            responseType: TxChallengeResponse.self
        )
    }

    func submitTransaction(request: TxSubmitRequest) async throws -> TxSubmitResponse {
        return try await client.post(
            url: "\(baseURL)/api/transaction/submit",
            body: request,
            responseType: TxSubmitResponse.self
        )
    }
}
